design: Service APIs implementation

youngnick opened the pull request:
This is a WIP of the Service APIs implementation. It's been slow going, as I've been trying to check as I went that I've got the Service APIs intent correct, and that the way we'll implement is relatively consistent with Contour's current state.

Interested mostly in feedback on the high-level parts, did I miss anything?

There's quite a bit more detail required, particular for HTTPRoute and TLSRoute.

Very interested to hear about any other areas you all think I should cover as part of this design.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+
+In terms of its stated goals, Contour is aiming at being an ingress controller - that is, a Layer 7 proxy with some api gateway functions.
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
```
I'm still confused on the practically of this, you're proposing that we recreate all the same logic to just watch for these two types (e.g. TCP / UDP)?

youngnick replied:
I'd guess that as part of this, we'd spin some code out into libraries, probably.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+Contour expects the `spec.addresses` field to be empty.
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible, as will listeners across Gateways.
```
Contour only supports two listeners today (e.g. http/https). Does this statement assume that you are permitted to create more dynamically?

youngnick replied:
The insecure/secure split definitely needs some thinking as a result of this. The initial  thought is that we would replicate the current behavior, which would mean accepting only one http and one https listener, but this will be tricky to do. We may need to do something different with automatic http->https redirect for the Service APIs.

youngnick replied:
I've opened #3263 with a new design for this, and will update this document accordingly.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+- Conflicts within a Gateway will result in the relevant Listeners both being rejected.
+- Conflicts between Gateways will result in the oldest Gateway (by creation timestamp) being kept.
+- The rejected Listener will have its status updated with the reason and the name of the conflicting Gateway.
+- Listeners are considered conflicting if all the fields out of `hostname`, `port`, `protocol`, and `tls` match.
```
Could TLS listeners also be in conflict if they reference different certificates?

sunjayBhatia replied:
relevant: https://github.com/projectcontour/contour/pull/3213/files#r552799600

youngnick replied:
I think I've covered this in a recent update, please tell me if it's clearer now.

stevesloka reviewed (commented):
What also do we do with things like:
- web sockets
- timeouts
- wildcard domain names (from api spec notes)
- regex headers
- exact path match type (also needed in ingress v1 support)

danehans commented:
xref: https://github.com/projectcontour/contour-operator/pull/171

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+
+### Contour Gateway model
+
+Contour considers a GatewayClass and associated set of Gateways to describe a single Envoy deployment, deployed and managed by something outside of itself.
```
can you remind me of the scenario where a Contour user would have more than one `Gateway`?

sunjayBhatia replied:
Looks like we have significant conflicts between the operator proposal and this one: https://github.com/projectcontour/contour-operator/pull/171/files#diff-1cbda5ba0844ea1d442d220ece0cd2f53066104e33808dffb0b7d8df6c2ff5d8R106

sunjayBhatia replied:
cc @danehans

sunjayBhatia replied:
https://github.com/projectcontour/contour-operator/pull/171/files#diff-1cbda5ba0844ea1d442d220ece0cd2f53066104e33808dffb0b7d8df6c2ff5d8R114

youngnick replied:
As discussed in our maintainer meeting, Contour will move to being configured to check a *specific*, named by namespace and name Gateway only. This is because of this line. I'll update this to make it clearer.

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+Contour can change this value to any `projectcontour.io/<value>` with a config setting (similar to how `ingress.class` works today.)
+Importantly, this means that changing this values requires a Contour restart; this means that the GatewayClass watcher can have this value baked in at runtime.
+
+If Contour finds more than one matching GatewayClass, this is a conflict, and Contour will use the oldest GatewayClass (by creation timestamp).
```
it looks like the "oldest object wins" strategy is used in a few places - are there any alternatives? why did you choose this strategy?

sunjayBhatia replied:
seems to be some mention of oldest-wins here as well: https://kubernetes-sigs.github.io/service-apis/httproute/#tls (notes section)

skriss replied:
just found https://kubernetes-sigs.github.io/service-apis/guidelines/#conflicts as well which explicitly mentions this as a principle

youngnick replied:
Yes, this is why I picked this, I think that oldest object wins is about the best we can hope for here.

sunjayBhatia commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+- Routes describe how traffic coming via the Gateway maps to the Services.
+
+In Contour, we've previously solved a lot of the same problems with HTTPProxy (and IngressRoute before it).
+That functionality can be described in the Service APIs by HTTPRoutes, HTTPSRoutes and TLSRoutes, as they describe Layer 7 ingress, as Contour does today.
```
doesn't look like it: https://kubernetes-sigs.github.io/service-apis/httproute/#spec (TLS optional)

youngnick replied:
🤦 Yeah, that should just be `HTTPRoutes` and `TLSRoutes`, for terminated and passthrough traffic respectively.

danehans commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+
+#### GatewayClass
+
+Contour will look for entries with `spec.controller` equal to `projectcontour.io/contour` by default.
```
@youngnick if contour operator is responsible for managing `GatewayClass`, then I would expect the value to be `projectcontour.io/contour-operator`. See https://github.com/projectcontour/contour-operator/pull/171 for additional details.

youngnick replied:
As discussed in our maintainer meeting, the controller default will be `projectcontour.io/contour`, which will be overridden by the operator to `projectcontour.io/contour-operator`. This is because the Operator is one way to install Contour and Envoy and an associated Gateway, but not the only one.

I'll clarify the properties that Contour expects a Gateway to have.

danehans commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+#### Gateway
+Contour will watch for entries with known GatewayClass name/namespace details, from the previous watch.
+
+Contour is not able to specify the address that the Envoy deployment will listen on, and so having any entries in Gateway's `spec.addresses` stanza will result in Contour marking those addresses as invalid in the Gateway's status.
```
The contour-operator will attempt to set these addresses and surface a status condition accordingly. See https://github.com/projectcontour/contour-operator/pull/171/files#r552860708 for add'l background.

danehans commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+Ideally the changes should be made in sequence so that the work required to implement this design can be done incrementally, possibly in parallel.
+
+## Alternatives Considered
+TODO: This will need further explanation of what the solution would look like if we didn't make one GatewayClass == One Contour.
```
The contour-operator [design doc](https://github.com/projectcontour/contour-operator/pull/171/files) more closely aligns with this alternative approach. An instance of Contour (i.e. Contour Deployment, Envoy DaemonSet, etc.) equals one Gateway. GatewayClass is simply a resource that points to an implementation-specific config object, i.e. `ContourConfig`. Contour will only be concerned with xRoute resources and using `xroute.spec.gateways[]` to know if it's responsible for handling the route.

youngnick replied:
I think we cleared this up in a maintainer's meeting, now that the Contour installation (= Contour deployment and Envoy DaemonSet and associated bits) is 1-to-1 with a named Gateway.

youngnick commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,229 @@
+# Service-APIs implementation design
+
+Status: Accepted
```
Well, I figured that if we merge it, it's accepted. Otherwise, this will get closed. Traditionally, we've just not updated this field, and everything is a draft, which is also not ideal. :)

skriss replied:
We could just drop the field, and assume anything that's been merged is accepted; anything still a PR is draft. Not a big deal though.

stevesloka replied:
The original idea was to submit ideas here without them being "accepted" yet, to gather thoughts, etc. We've been waiting to commit the design until it's accepted in a way, so removing the field might better follow how we're implementing features.

youngnick replied:
Yeah, removing the field may make more sense, if it's merged, it's accepted. If we change things like that, we'll need to:
- update the template and remove the field
- update CONTRIBUTING.md/ other policy docs
- update the old PRs to remove the field.

youngnick commented:
In the latest round of changes, I've updated a few things:
- change the model that Contour only looks at a single Gateway (since it's part of the implementation of that specific Gateway).
- Contour is configured with the full name (name and namespace) of the Gateway.
- Added references to #3263 and updated the design assuming it's going ahead. If not, this is still doable, just will require extra rules for Listener merging.

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,253 @@
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
```
I'm OK with either order, as long as we document it.

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,253 @@
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
```
Is it a goal to eventually sunset `HTTPProxy` and have service APIs be the only API for Contour? If not yet a goal, do we think it might be some day? Just curious how others are thinking about this, and whether we expect to support multiple APIs indefinitely or if we hope to converge on one.

youngnick replied:
I'd like to think that, if the Service APIs had full feature parity with HTTPProxy, and our users were willing to move, we would remove the CRD.

I don't think either of those is likely though, so what I'm trying to do is behave as though that might be the case someday. Practically, I don't think it will be, but I'm not convinced of the benefit of having our own CRD if the above two conditions are fulfilled.

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,253 @@
+Contour will only update the `status` of Gateway objects.
+
+### HTTPRoute
+Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in a valid Gateway `spec.listeners[].routes`,
```
```suggestion
Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
```

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,253 @@
+The output of this watcher is Envoy configuration.
+
+### TLSRoute
+Contour will watch all TLSRoute objects, and filter for entries matching a label selector in a valid Gateway `spec.listeners[].routes`,
```
```suggestion
Contour will watch all TLSRoute objects, and filter for entries matching a label selector in its configured Gateway's`spec.listeners[].routes`,
```

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,253 @@
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+## Goals
```
is it a goal for Contour's service-apis support to reach ~feature parity with `HTTPProxy`? If so, how will we implement features that aren't explicitly part of the service-apis spec? And if not, what set of features _do_ we expect to implement?

Another way of asking this is, are we going to use annotations to implement features that aren't directly a part of service-apis, like we do with Ingress?

youngnick replied:
I'll put something in the background to summarise this, but:
- I think the goal should absolutely be that eventually service-apis has feature parity with HTTPProxy
- the upstream does not have support for many features right now
- we should implement what is implementable right now
- we should collaborate with upstream to work out what features are part of the core, extension, and implementation specific sections of service-apis.

The service-apis are specifically built with extension points so you don't have to use annotations. That's what the reference points that allow you to point to any object, but default to a specific kind, are for.

youngnick commented:
To answer @stevesloka's question about additional features, I've also put some more in the documentation about what parts of the API we will target, and how we will build support.

In short, we'll build whatever is in v1alpha1 in the initial cut, then work with the upstream to define the other features (timeouts, websockets, regex headers etc) in the upstream, and support that.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,258 @@
+This design is intended to cover the initial `v1alpha1` release of the Service APIs. We will aim to implement the core featureset of the APIs at that point.
+We will then work with the upstream community on features that Contour and HTTPProxy currently support, but the Service APIs do not, and how best to represent those features in the APIs.
+So, there are some features that are currently gaps for Contour (wildcard domain names, and exact path matching seem like big ones),
+and some that Contour supports that the Service APIs do not (webscokets, configurable timeouts, header replacement, external auth, rate limiting, and so on).
```
nit: s/webscokets/websockets/

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,258 @@
+
+The output of this watcher is Envoy configuration.
+
+## Detailed Design
```
do we want to have one DAG processor for all service-apis objects? One per object type (e.g. `GatewayProcessor`, `HTTPRouteProcessor`, etc)? Something else?

youngnick replied:
I'm not sure. I think that splitting the processors up would be a bit different to how we've done things before, but that's not necessarily bad. It's probably worth the implementor giving this a go.

skriss replied:
:+1:

skriss reviewed (commented):
I don't have any other significant feedback here, I think this points us in the right direction. I'd be OK working out the remaining details of processing objects in code, especially given the alpha status of the API.

stevesloka commented:
I don't have any additional large comments. There are a lot of behavior details in the API spec which we'll also need to reference during implementation. 

There are still quite a few TODO's throughout the doc that will need another review pass from folks if they are expaned upon. 

Thanks for this work @youngnick!

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,258 @@
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
```
Do you see this as new flags to Contour or placed into the config file?

youngnick replied:
I think config file for this one, I'll update.

youngnick commented:
I've updated this document a bit today, but I still have a few TODOs. I've removed the WIP as this is much closer now, but left it as draft as it will still need some more review passes.

youngnick commented:
Okay, I've marked this ready for review, I've finished up all the outstanding TODOs, I think. Can I get some more eyes across this, then we can get it merged.

sunjayBhatia commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,326 @@
+
+Contour allows the configuration of arbitrary ports, **as long as those ports are exposed via the Envoy service**.
+The port that is configured for the listener is the port as it appears at the Service.
+That is, the Envoy Service `port` number is what the Listener must specify, regardless of if there is a `targetPort` set on the Service.
```
Maybe a comment somewhere making it explicit that we will start watching the Envoy service would be good to add. Also it forces users (cluster operators?) to deploy their Envoys in a certain way. Practically speaking its unlikely someone had their Envoys deployed much differently than we have in the examples or the operator, but it is technically a new constraint I guess? Maybe not that serious

youngnick replied:
I've updated this quite a bit to reflect the fact that we're not going to allow arbitrary ports for now. There's definitely a problem that will need to be solved with some experimentation during implementation here.

sunjayBhatia commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,326 @@
+- Deprecate `--expermental-service-apis` command-line flag, so that it does nothing but log a referral to the new config item.
+- Implement the Service-APIs processor as a single DAG processor.
+- Release a version with Service-APIs support still marked experimental.
+- Remove the experimental tag once we're happy with the support.
```
Might be worth adding some notes about how before this stage we will try to evaluate our implementation and work with upstream to work out issues, conformance testing, etc.

sunjayBhatia reviewed (commented):
Overall looks reasonable to me, theres lots of details to get right but with incremental PRs and a good focus in PR review on test coverage we should be good to refer to this design for help

tsaarni commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,342 @@
+Contour will only update the `status` of Gateway objects.
+
+### HTTPRoute
+Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
```
Did I understand it correctly that `HTTPRoute` and `TLSRoute` resources are first selected from Gateway's own namespace or from namespaces listed in Gateway's `spec.listeners[].routes.namespaces` (if given), and after that optional label selector `spec.listeners[].routes.selector` is applied (if given).  With the current wording this two level process might be bit unclear. 
 
Since the selection logic is more `Gateway` topic, it could be moved from HTTPRoute and TLSRoute chapters to Gateway chapter, but it is readable as it is today as well.

By the way, the new "inclusion" logic seems to solve #2206.

youngnick replied:
Yes, you're correct @tsaarni. I'll try and update this to be closer to the spec.

Also, yes, I know that #2206 is partially solved by HTTPRoutes, although they currently cannot be included in themselves (so you can only do the equivalent of one layer of HTTPProxy includes). There was a vision of doing that initially for the service-apis, but I'm not sure where it's at, and it would add significant complexity to processing of the HTTPRoute objects.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,342 @@
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
+This will be a configuration file parameter.
+When a Gateway is not supplied, Contour will not do Service-APIs processing, or in other words, the existing `--experimental-service-apis` flag will be obsoleted and removed).
```
+1 on removing the flag. 

Should we default the name/namespace of a gateway (i.e. contour/projectcontour) so given a generic deployment, someone could use that as a default and not require a config file change?

Contour can still check if the service-apis CRDs exist in the cluster and use that as a key for setting up informers.  I just merged https://github.com/projectcontour/contour/pull/3283 which acts this way, but is different than this design doc, but @skriss asked which is a good point.

youngnick replied:
So, the flow if you want to use the service-apis is this:
- install the objects in the cluster
- create a Gateway for Contour to reference
- Configure Contour to use that

If you don't do all those three things, then no service-apis.

That does make sense, although it means that if you install the objects and don't change the default, we should end up logging something saying "hey, you asked me to look for a Gateway called `projectcontour/contour`, but it doesn't exist". Not a huge deal.

stevesloka replied:
Right now I have a default programmed, but we could remove that, just thought it might be nice to have some way of allowing someone to create a Gateway without needing to reconfigure / restart Contour. 

This came up in a review with @skriss and thought it was a good point that what I PR'd conflicted with this design and thought we should clear up. =)

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,342 @@
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with its named Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
```
If we did watch the class, we could allow a single gateway in a cluster to be configured by a single contour without any configuration file changes. But when a second Gateway for the same class controller was created you'd need to specify.

Would just be nice to allow users to configure Gateways without having to edit the configmap + restart contour instances.

youngnick replied:
I think that it's better to be explicit for now. We can revisit this later if the configuration proves too complicated. This way we can avoid a more complicated state like we had for the ingress.class annotation, where the behavior if you don't specify a value is not straightforward.

stevesloka commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,342 @@
+
+Contour will only ever update the `status` of HTTPRoute objects.
+
+Errors or conflicts here will render that section of the config invalid.
```
For a missing `serviceName` or `LocalObjectReference`, the docs say: 
```
If the referent cannot be found, the route must be dropped from the Gateway. The controller should raise the “ResolvedRefs” condition on the Gateway with the “DegradedRoutes” reason. The gateway status for this route should be updated with a condition that describes the error more specifically.
```

Does that mean the entire `HTTPRoute` is invalid or just the specific `HTTPRoute.HTTPRouteRule` is invalid?

sunjayBhatia replied:
Yeah I wasn't sure about that either (and how it would play with the proposal you had around keeping as much routability as possible even in partially applied resources), I think we chatted about this in office hours or something but I'm not sure there was a conclusion

youngnick replied:
You're right, I think we need to implement this per the spec for now, and raise this upstream as a concern.

stevesloka replied:
Chatting in Slack the thought is that only that HTTPRoute.Rule is invalidated so I think the design is correct unless we want to specifically call it out better: https://kubernetes.slack.com/archives/CR0H13KGA/p1612463856030500?thread_ts=1612459180.030400&cid=CR0H13KGA

youngnick replied:
Thanks for asking this question in the upstream Slack @stevesloka. I agree with Harry's answer there - we will implement this as "conflicts will invalidate that *rule*" not "conflicts with invalidate the whole Route" for now, and see if we can push that upstream.

skriss commented on design/service-apis-implementation.md:
```diff
@@ -0,0 +1,342 @@
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
```
Some aspects of this are not entirely clear to me, e.g. what if there's a service-apis Listener with a wildcard domain name that overlaps with an HTTPProxy with a specific domain name, or what if some other service-apis Listener settings conflict with Contour's listener settings -- but I'm also not sure it's that important right now, and we can figure it out during implementation.

youngnick replied:
I think the wildcard domain name thing will be sorted out by implementing wildcard domains everywhere - we should do that before starting with this, since we need to do it for Ingress v1 anyway. I think the limited form of wildcard FQDN defined in the service-apis is probably implementable in HTTPProxy as well.

I think the LIstener settings details should be sorted out during implementation, but I think that the service-apis spec should probably override the Contour config one, since it's more specific (I think). I don't think it matters really, as long as we document it.

skriss reviewed (approved):
no further comments/questions from me, I think we can sort out remaining details during implementation. 🎉

stevesloka reviewed (approved):
I think this is good enough to start implementation and dig into what comes out of that when we hit those bits. 🎉

youngnick commented:
Looks like the new ExternalName integration test is failing on `main`, I'm going to merge this design, and take a look.
